### ISSUE TITLE
Strange behavior during learning

### ISSUE THREAD

**Issue by Rodnyy** [Unity-Technologies/ml-agents#3150]
Hi, I am working on training an agent to play a game called mastermind. In a short way, the agent is guessing a secret code that consists of 5 pieces with 6 colors to choose from and after each guess he receives feedback. I started with a simplified version with just 3 colors which went quite well but when I try to train in on the more complicated one I always fail. I tried to fiddle with the parameters accordingly to the documentation but with no success. So I decided that probably the complexity of the task needs a curriculum learning and a longer learning period so I tried to learn it overnight but then something strange happened. The first two parts of curriculum learning went pretty swiftly even though instead on 0,1 reward it switched the parameters on -0.1 and it did the same for the second time. I do not if I did something wrong but can't explain this. The main problem came later during the training because the mean reward was steadily increasing but suddenly dropped to the bottom and never came back. Also, it's worth mentioning that I used SAC for learning. Is it possible to say what might have caused such a behavior? Is it possible that the task is too complex or do I need better parameter settings?

![MLagents2](https://user-images.githubusercontent.com/33638523/71715343-598dac00-2e11-11ea-801a-861f6d24974b.png)
![MLagents3](https://user-images.githubusercontent.com/33638523/71715344-5a264280-2e11-11ea-8628-59dc25f09f30.png)
![MLagents](https://user-images.githubusercontent.com/33638523/71715347-5a264280-2e11-11ea-87a9-70c76af197c2.png)

``default:
    trainer: sac
    batch_size: 64
    buffer_size: 50000
    buffer_init_steps: 2000
    hidden_units: 512
    init_entcoef: 0.25
    learning_rate: 1e-4
    learning_rate_schedule: constant
    max_steps: 5.0e10
    memory_size: 256
    normalize: true
    num_update: 1
    train_interval: 1
    num_layers: 3
    time_horizon: 16
    sequence_length: 64
    summary_freq: 1000
    tau: 0.005
    use_recurrent: false
    vis_encode_type: simple
    reward_signals:
        extrinsic:
            strength: 1.0
            gamma: 0.995
``

**Comment by Rodnyy** [comment]
I have changed some of the parameters and tried to train it again but unfortunately, it performed even worse the same sudden extremely bad behavior after a certain amount of steps. Can anybody point me on how to improve the learning of more complicated behaviors? What is the best practice of changing the parameters? How do I recognize what to change and how?

**Comment by niskander** [comment]
Did you try saving and loading the model after the sharp drop to see how it's behaving? It's possible that it ran into a bug in the code or got stuck doing something.

**Comment by Rodnyy** [comment]
Hi, thanks for the answer, yes I have tried loading the model and training again for a while but without success. I have noticed that there is something called checkpoints during the learning but I am not sure if it's possible to start the learning from some checkpoint before the drop. If it was possible I think it might it might help.

**Comment by oxters168** [comment]
I know it's almost been two months since the last comment, but I wanted to pitch in. I'm having a similar issue with my agents and I'm still trying to figure out exactly what is causing them to not learn properly. Anyways I've had to go back checkpoints and the way I'm doing it is by going into where the checkpoints are saved deep in the models folder. I open the checkpoint file with notepad, delete the checkpoints until the one I want and change the pointer to the one I want. Then I delete the checkpoint files in the folder that I don't need. Then I resume training with the --load parameter and once the Unity window opens up I end the training to build the new nn model. Of course before all this I backup the models and summary folder just to be safe.
